src/bst.py

```diff
@@ -1,32 +1,219 @@
+                self.right.pivot_left()
+                # self.right.pivot_rl()
+
         elif nn.value < self.value:
             if self.left:
                 self.left.insert(nn)
             else:
                 self.left = nn
-            self.depth = max(self.depth, self.left.depth + 1)
+            self._depth = max(self._depth, self.left._depth + 1)
+
+            if self.balance() >= 2 and self.left.balance() > 0:
+                self.left.pivot_right()
+            if self.balance() >= 2 and self.left.balance() < 0:
+                self.left.right.pivot_left()
+                self.left.pivot_right()
+                # self.left.pivot_lr()
+
+    @property
+    def depth(self):
+        try:
+            left_depth = self.left._depth
+        except AttributeError:
+            left_depth = 0
+        try:
+            right_depth = self.right._depth
+        except AttributeError:
+            right_depth = 0
+
+        return max(left_depth, right_depth) + 1
+
+    def balance(self):
+        try:
+            left_depth = self.left._depth
+        except AttributeError:
+            left_depth = 0
+        try:
+            right_depth = self.right._depth
+        except AttributeError:
+            right_depth = 0
+
+        return left_depth - right_depth
+
+    def in_order(self):
+        '''recursive in order traversal'''
+        if self.left:
+            for item in self.left.in_order():
+                yield item
+        yield self.value
+        if self.right:
+            for item in self.right.in_order():
+                yield item
+
+    def post_order(self):
+        '''recursive post order traversal'''
+        if self.left:
+            for item in self.left.post_order():
+                yield item
+        if self.right:
+            for item in self.right.post_order():
+                yield item
+        yield self.value
+
+    def pivot_right(self):
+        """Perform a right rotation on the node."""
+        pivot = self
+        temp = self.parent
+        sib = self._right
+
+        if temp.parent and temp.parent < temp:
+            temp.parent.right = self
+        elif temp.parent:
+            temp.parent.left = self
+        self.parent = temp.parent
+        self._right = temp
+        temp.parent = self
+        temp._left = sib
+
+        if temp.right and temp.left:
+            temp._depth = max(temp.right.depth, temp.left.depth) + 1
+        elif temp.right:
+            temp._depth = temp.right.depth + 1
+        elif temp.left:
+            temp._depth = temp.left.depth + 1
+        else:
+            temp._depth = 1
+
+        if pivot.left:
+            pivot._depth = max(pivot.right.depth, pivot.left.depth) + 1
+        else:
+            pivot._depth = pivot.right.depth + 1
+
+    def pivot_left(self):
+        """Perform a left rotation on the node."""
+        pivot = self
+        temp = self.parent
+        sib = self._left
+
+        if temp.parent and temp.parent > temp:
```

Comment:
There are many scenarios when temp is None, and so this will cause an AttributeError here.